### ISSUE TITLE
Crash with weird self types

### ISSUE THREAD

**Issue by stedolan** [ocaml/ocaml#9885]
The following (entirely reasonable and normal...) code causes assertion failures in typeclass / ctype:
```
class bad x = object (s : 's) val l : 's list = x end
let bs : bad list = []
class b = object inherit bad bs end
let a = object inherit bad bs end
```
(Lines 3 and 4 cause two different errors)

My guess is that there should be a "Self type cannot escape its class" error about the parameter `x`: I think its inferred type is ill-formed, referring to a self type from outside the class.

**Comment by MattX** [comment]
I think this is a duplicate of the (poorly named) #7894. The regression was introduced in 86e1c8a (merged as part of #1620).

**Comment by stedolan** [comment]
This is indeed fixed in trunk.
